### PR TITLE
feat: MoE and Mamba support for QARL 

### PR DESCRIPTION
**Pull request by mxinO**
# What does this PR do ?

* Adding MoE and Mamba support for QARL
* Updating modelopt to latest
* Fix potential issues for vllm fake calibration 

## Example training
Nemotron Nano3 NVFP4 W4A4 default quantization config, QA-OPD, KL loss converge curve, Top-k = 300
Screenshot 2026-06-23 at 15 57 57

# Issues

# Usage
* **You can potentially add a usage example below**

```python
# Add a code snippet demonstrating how to use this
```

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](/NVIDIA-NeMo/RL/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you run the unit tests and functional tests locally? Visit our [Testing Guide](/NVIDIA-NeMo/RL/blob/main/docs/testing.md) for how to run tests
- [x] Did you add or update any necessary documentation? Visit our [Document Development Guide](/NVIDIA-NeMo/RL/blob/main/docs/documentation.md) for how to write, build and test the docs.

# Additional Information
* The PR series for support this, 
   * https://github.com/NVIDIA-NeMo/Megatron-Bridge/pull/3816
   * https://github.com/NVIDIA-NeMo/RL/pull/2759
   * https://github.com/NVIDIA-NeMo/RL/pull/2793
 * TODOs
 - [x] the new megatron updated the implementation of conv1d, so currently conv1d quantizer doesn't work, but we also don't quantize it usually, but we will need a fix. And this will cause the fail when exporting the quantized nano3 ckpt. This can be fixed later with a small pr. Draft pr: https://github.com/NVIDIA/Model-Optimizer/pull/1716

### CODE REVIEW

**Review by ashors1 — DISMISSED** [review]
LGTM, thank you Meng!

**Review by zhangyimi — DISMISSED** [review]
LGTM

**Review by ashors1 — DISMISSED** [review]

File: nemo_rl/algorithms/distillation.py [hunk 1]
```diff
@@ -986,13 +1020,30 @@ def validate(
...
+                    tokenizer=tokenizer,
+                    task_to_env=val_task_to_env,
+                    max_seq_len=None,
+                    generation_config=generation_config,
+                    max_rollout_turns=None,
+                    greedy=False,
+                )
+                val_batch = nemo_gym_rollout_result.final_batch
+                gen_metrics = nemo_gym_rollout_result.rollout_metrics
+                additional_metrics_to_report = gen_metrics
```
Comment (terrykong) [inline nemo_rl/algorithms/distillation.py]: [`nemo_rl/algorithms/distillation.py:1044`](https://github.com/NVIDIA-NeMo/RL/blob/13d1c20c333158e13049ca791b6bdf5797b0066a/nemo_rl/algorithms/distillation.py#L1044)

`validate()` merges NeMo-Gym rollout metrics into `val_metrics` without applying the [`_should_log_nemo_gym_responses` filter](https://github.com/NVIDIA-NeMo/RL/blob/13d1c20c333158e13049ca791b6bdf5797b0066a/nemo_rl/algorithms/distillation.py#L660) that the training loop uses. The large `full_result` Tables will always be emitted during validation regardless of the opt-in setting.

Additionally, `additional_metrics_to_report` is reassigned (`=`) rather than accumulated per batch — with >1 val batch only the last batch's metrics would survive. (Latent today since `val_batch_size == max_val_samples` forces 1 batch.)

Consider applying the same filter here:

```suggestion
                val_batch = nemo_gym_rollout_result.final_batch
                gen_metrics = nemo_gym_rollout_result.rollout_metrics
                if not _should_log_nemo_gym_responses(master_config):
                    for key in list(gen_metrics):
                        if "full_result" in key:
                            gen_metrics.pop(key)
                additional_metrics_to_report = gen_metrics
```

File: tests/test_suites/llm/grpo-qwen3-30ba3b-4n4g-megatron-qa-nvfp4.sh [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+MAX_STEPS=1
+NUM_RUNS=$(( (MAX_STEPS + STEPS_PER_RUN - 1) / STEPS_PER_RUN ))  # Round up
+NUM_MINUTES=45
+# ===== END CONFIG =====
+
+exit_if_max_steps_reached
+
+# Run the experiment
+cd $PROJECT_ROOT
+export DISABLE_MODELOPT_LAYER_SPEC=1
```
Comment (terrykong) [inline tests/test_suites/llm/grpo-qwen3-30ba3b-4n4g-megatron-qa-nvfp4.sh]: [`grpo-qwen3-30ba3b-4n4g-megatron-qa-nvfp4.sh:22`](https://github.com/NVIDIA-NeMo/RL/blob/13d1c20c333158e13049ca791b6bdf5797b0066a/tests/test_suites/llm/grpo-qwen3-30ba3b-4n4g-megatron-qa-nvfp4.sh#L22)

This nightly sets `DISABLE_MODELOPT_LAYER_SPEC=1`, routing to plain Megatron layer specs and bypassing the new [`modelopt_mamba_stack_spec`](https://github.com/NVIDIA-NeMo/RL/blob/13d1c20c333158e13049ca791b6bdf5797b0066a/nemo_rl/modelopt/models/policy/workers/utils.py#L190) / [`partial(get_gpt_modelopt_spec,...)`](https://github.com/NVIDIA-NeMo/RL/blob/13d1c20c333158e13049ca791b6bdf5797b0066a/nemo_rl/modelopt/models/policy/workers/utils.py#L179) default path.

The GPT modelopt-spec path IS covered by the existing `qa_distillation_megatron.sh` functional test (no DISABLE flag). However the **Mamba/hybrid** modelopt-spec path (`modelopt_mamba_stack_spec` on Nano3) has no automated coverage — only manual smoke testing.

Could you consider adding a Nano3 nightly that exercises the default modelopt spec (without the DISABLE flag)?

File: docs/guides/quantization-aware-rl.md [hunk 3]
```diff
@@ -17,9 +17,19 @@ The following workflow + quantization recipe combinations have been validated en
 | QA-Distillation | W4A4 | `NVFP4_DEFAULT_CFG` (NVFP4 weights + NVFP4 activations) | ✅ Converges | `examples/modelopt/qa_distillation_math_megatron.yaml` |
 | QA-GRPO | W4A16 | `examples/modelopt/quant_configs/nvfp4_a16.yaml` (NVFP4 weights, native-dtype activations) | ✅ Converges | `examples/modelopt/qa_grpo_llama8b_megatron.v2.yaml` |
 | QA-GRPO | W4A4 | `NVFP4_DEFAULT_CFG` | ⚠️ Known convergence issue | `examples/modelopt/qa_grpo_math_megatron.yaml` |
+| QA-Distillation | W4A4 | `examples/modelopt/quant_configs/nano3_nvfp4_default.yaml` | ✅ Smoke tested on hybrid MoE/Mamba | `examples/modelopt/qa_distillation_nano3_megatron.yaml` |
```
Comment (terrykong) [inline docs/guides/quantization-aware-rl.md]: [`quantization-aware-rl.md:20`](https://github.com/NVIDIA-NeMo/RL/blob/13d1c20c333158e13049ca791b6bdf5797b0066a/docs/guides/quantization-aware-rl.md#L20)

The honest “Smoke tested” labeling here is appreciated — it clearly sets user expectations.

When you have the data available, it would be helpful to share: (1) convergence curves for at least one MoE and one Mamba/hybrid config, (2) whether W4A4 on Nano3 hybrid actually converges or only runs without crashing (given that [dense W4A4 GRPO has a known convergence issue](https://github.com/NVIDIA-NeMo/RL/blob/13d1c20c333158e13049ca791b6bdf5797b0066a/docs/guides/quantization-aware-rl.md#L19)), and (3) throughput/memory vs BF16 baseline for the Qwen3-30ba3b MoE path. Not blocking — just a follow-up ask.

File: nemo_rl/modelopt/models/generation/vllm_quant_patch.py [hunk 4]
```diff
@@ -14,17 +14,62 @@
...
+
+    Nonfinite dummy activations are sanitized before calibration reduce. The
+    patch is active only inside the dummy-weight prolog, before runtime
+    generation starts and before real amax values are loaded.
+    """
+    _original_collect = MaxCalibrator.collect
+
+    @torch.no_grad()
+    def _safe_collect(self, x):
+        if x.numel() > 0 and not torch.isfinite(x).all():
```
Comment (terrykong) [inline nemo_rl/modelopt/models/generation/vllm_quant_patch.py]: [`vllm_quant_patch.py:62`](https://github.com/NVIDIA-NeMo/RL/blob/13d1c20c333158e13049ca791b6bdf5797b0066a/nemo_rl/modelopt/models/generation/vllm_quant_patch.py#L62)

Nit (defensive hardening): `torch.isfinite(x).all().__bool__()` would crash on meta-device tensors, whereas the original `MaxCalibrator.collect` handles them with an early return. In practice the prolog operates on materialized dummy CUDA weights so no meta tensor reaches `collect` here — but adding a guard preserves the upstream contract:

```suggestion
    @torch.no_grad()
    def _safe_collect(self, x):
        if x.device.type != "meta" and x.numel() > 0 and not torch.isfinite(x).all():
            x = torch.nan_to_num(x, nan=0.0, posinf=0.0, neginf=0.0)
        _original_collect(self, x)
```

File: nemo_rl/algorithms/distillation.py [hunk 5]
```diff
@@ -84,7 +90,7 @@ class DistillationConfig(TypedDict):
     # Whether to run validation on the last training step. Setting this to True ensures the
     # final checkpoint has validation metrics, which is required for get_best_checkpoint_path().
     val_at_end: bool
-    max_val_samples: int
+    max_val_samples: int | None  # None for NeMo-Gym compatibility
```
Comment (terrykong) [inline nemo_rl/algorithms/distillation.py:104]: [`distillation.py:93`](https://github.com/NVIDIA-NeMo/RL/blob/13d1c20c333158e13049ca791b6bdf5797b0066a/nemo_rl/algorithms/distillation.py#L93)

Informational: widening to `int | None` means that if a user sets `max_val_samples: null` in a non-NeMo-Gym config, [`validate()` at line 1019](https://github.com/NVIDIA-NeMo/RL/blob/13d1c20c333158e13049ca791b6bdf5797b0066a/nemo_rl/algorithms/distillation.py#L1019) would hit a TypeError (`None + int`). The NeMo-Gym runner [overrides null → len(val_dataset)](https://github.com/NVIDIA-NeMo/RL/blob/13d1c20c333158e13049ca791b6bdf5797b0066a/examples/nemo_gym/run_distillation_nemo_gym.py#L131) before `validate()`, and the default YAML sets 512, so this is latent. Same pattern exists in grpo.py — not a regression, just noting for awareness.

File: nemo_rl/algorithms/distillation.py [hunk 6]
```diff
@@ -536,7 +542,10 @@ def distillation_train(
         student_generation = student_policy  # type: ignore
         NEED_REFIT = False
     POLICY_GENERATION_STALE = True  # tracks if generation needs a refit before running
-    assert student_generation is not None
+    assert student_generation is not None  # for mypy type check
+    use_nemo_gym = _should_use_nemo_gym(master_config)
```
Comment (terrykong) [inline nemo_rl/algorithms/distillation.py:666]: i kind of feel this PR should be broken up into nemo-gym in distillation and MoE/Mamba in QARL. is there some reason these have to be bundled together?

@RayenTian @yfw please opine

File: nemo_rl/modelopt/models/generation/vllm_quant_patch.py [hunk 7]
```diff
@@ -14,17 +14,62 @@
...
+
+    The dummy-calibration amax is *meant* to be discarded — the prolog
+    sentinels every enabled quantizer's `_amax` to `-1.0` immediately
+    afterwards, and Megatron's real amax is loaded via
+    `vllm_quant_backend.input_amax_loader` during refit (`max(-1.0,
+    real)=real`). So a fully-NaN input here should produce zero amax
+    rather than crash the prolog.
+
+    Scoping this monkey-patch to the prolog (instead of editing
+    `MaxCalibrator.collect` in modelopt) keeps modelopt's source pristine
```
Comment (terrykong) [inline nemo_rl/modelopt/models/generation/vllm_quant_patch.py:49]: can we push this into modelopt? this is a little bit of a code smell

File: nemo_rl/modelopt/models/policy/workers/utils.py [hunk 8]
```diff
@@ -165,16 +169,26 @@ def get_modelopt_checkpoint_dir() -> str:
...
         local_core_attention=False,
         remap_te_layernorm=True,
         real_quant_cfg="None",
         use_arbitrary_attention_mask=False,
     )
+
+
+def get_quantization_mamba_stack_spec():
+    """Return a checkpoint-serializable Mamba stack-spec callback."""
+    if int(os.environ.get("DISABLE_MODELOPT_LAYER_SPEC", "0")):
```
Comment (terrykong) [inline nemo_rl/modelopt/models/policy/workers/utils.py]: can we please make this a config value? this way the args someone runs with is documented in wandb

**Review by terrykong — COMMENTED** [review]
Reviewed by 5 specialized agents (rl-expert, bug-finder, test-agent, comment-reviewer, devil-advocate).

**Overall**: Solid feature PR adding MoE and Mamba support for QARL with NeMo-Gym distillation integration. The code is well-structured and the honest "Smoke tested" labeling in docs is appreciated. Main suggestions: (1) apply the `full_result` logging filter in validation too, (2) consider automated test coverage for the Mamba/hybrid modelopt-spec path, and (3) share convergence/performance evidence when available.

3 suggestions, 2 informational notes. No blocking issues.

_Generated by Claude Code_

File: nemo_rl/algorithms/distillation.py (hunk 6 above)
Comment (mxinO) [reply nemo_rl/algorithms/distillation.py:666]: Yes, that's more reasonable, I added to this PR because nano3's RL is using nemo gym, and I'm adding QA distillation for it which also needs nemo gym. But I can split to two PRs, if you prefer.

File: nemo_rl/modelopt/models/generation/vllm_quant_patch.py (hunk 7 above)
Comment (mxinO) [reply nemo_rl/modelopt/models/generation/vllm_quant_patch.py:49]: This is a special case, the calibration here is a dummy calibration for init amax, and the weight here is also dummy, which leads to nan values easily, we need tolerate the nan values here, but it is not needed in modelopt main branch which raises errors for nan values.

File: nemo_rl/modelopt/models/policy/workers/utils.py (hunk 8 above)
Comment (mxinO) [reply nemo_rl/modelopt/models/policy/workers/utils.py]: yeah, good suggestion, I moved it to config.

File: docs/guides/quantization-aware-rl.md (hunk 3 above)
Comment (mxinO) [reply docs/guides/quantization-aware-rl.md]: Thanks for pointing out, the distillation always converges, I have runs with nano3, I will update the wording here.

File: nemo_rl/algorithms/distillation.py (hunk 1 above)
Comment (mxinO) [reply nemo_rl/algorithms/distillation.py]: added the filter, and changed the `additional_metrics_to_report` to list and use `aggregate_rollout_metrics`.

File: tests/test_suites/llm/grpo-qwen3-30ba3b-4n4g-megatron-qa-nvfp4.sh (hunk 2 above)
Comment (mxinO) [reply tests/test_suites/llm/grpo-qwen3-30ba3b-4n4g-megatron-qa-nvfp4.sh]: Yeah, that's more safe, I wanted to avoid too many nightly tests. Added a test for nano3 with modelopt layer spec.

**Review by terrykong — COMMENTED** [review]
lgtm. @arnavk-nvidia to review/approve

@mxinO could you please include some convergence curves in the PR description? It helps to have some demonstration of features that affect convergence have some public demonstration of its efficacy. Also gives devs a reference to your run

File: nemo_rl/modelopt/models/generation/vllm_quant_worker.py [hunk 9]
```diff
@@ -88,12 +94,7 @@ def __init__(self, *args, **kwargs):
...
     def _create_engine(self, llm_kwargs: dict[str, Any]) -> None:
-        llm_kwargs["worker_cls"] = (
-            "nemo_rl.modelopt.models.generation.vllm_quant_patch.FakeQuantWorker"
-        )
-        llm_kwargs["worker_extension_cls"] = (
-            "nemo_rl.modelopt.models.generation.vllm_quant_backend.VllmQuantInternalWorkerExtension"
-        )
+        _configure_quant_vllm_kwargs(llm_kwargs)
         if self.cfg["quant_cfg"]:
             os.environ["VLLM_QUANT_CFG"] = self.cfg["quant_cfg"]
```
Comment (arnavk-nvidia) [inline nemo_rl/modelopt/models/generation/vllm_quant_worker.py:99]: Why switch to environment variables here?

File: nemo_rl/modelopt/models/generation/vllm_quant_worker.py [hunk 10]
```diff
@@ -36,12 +47,7 @@ def __init__(self, *args, **kwargs):
...
-        llm_kwargs["worker_cls"] = (
-            "nemo_rl.modelopt.models.generation.vllm_quant_patch.FakeQuantWorker"
-        )
-        llm_kwargs["worker_extension_cls"] = (
-            "nemo_rl.modelopt.models.generation.vllm_quant_backend.VllmQuantInternalWorkerExtension"
-        )
+        _configure_quant_vllm_kwargs(llm_kwargs)
         if self.cfg["quant_cfg"]:
             print("setting VLLM_QUANT_CFG to: ", self.cfg["quant_cfg"])
             os.environ["VLLM_QUANT_CFG"] = self.cfg["quant_cfg"]
```
Comment (arnavk-nvidia) [inline nemo_rl/modelopt/models/generation/vllm_quant_worker.py:53]: Why switch to environment variables here?

File: nemo_rl/modelopt/models/generation/vllm_quant_worker.py (hunk 9 above)
Comment (mxinO) [reply nemo_rl/modelopt/models/generation/vllm_quant_worker.py:99]: The qunat config in vllm side is passed by env variable, we can't directly add new args to the vllm LLM Kwargs.

**Review by arnavk-nvidia — APPROVED** [review]
